javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -151,67 +169,66 @@ private boolean arePointingToSameLeaf(int i, int j) {
     }
 
     /* drop root node while it has a single element */
-    private static <T> BitMappedTrie<T> collapsed(Object[] array, int offset, int length, int shift) {
+    private static <T> BitMappedTrie<T> collapsed(ArrayType<T> type, Object array, int offset, int length, int shift) {
         for (; shift > 0; shift -= BRANCHING_BASE) {
-            final int skippedElements = array.length - 1;
+            final int skippedElements = obj().lengthOf(array) - 1;
             if (skippedElements != digit(offset, shift)) {
                 break;
             }
-            array = getAt(array, skippedElements);
+            array = obj().getAt(array, skippedElements);
```

Comment:
We should not use (or even know about) `obj()` here (in BMT) at all. Instead use a static import of the ObjectArrayType methods `getAt` and `setAt` and use those here. We do not need the cast to `ArrayType<T>` because we deal with Object here.

(If we can't statically import them, we should use the qualified call ObjectArrayType.getAt...)

---

But it is still not clear to me why we are using ObjectArrayType directly. Can't the values be primitive?